scanpy/experimental/pp/_highly_variable_genes.py

```diff
@@ -0,0 +1,330 @@
+    else:
+        batch_info = adata.obs[batch_key].values
+    n_batches = len(np.unique(batch_info))
+
+    # Get pearson residuals for each batch separately
+    residual_gene_vars = []
+    for batch in np.unique(batch_info):
+
+        adata_subset = adata[batch_info == batch]
+
+        # Filter out zero genes
+        with settings.verbosity.override(Verbosity.error):
+            nonzero_genes = filter_genes(adata_subset, min_cells=1, inplace=False)[0]
+        adata_subset = adata_subset[:, nonzero_genes]
+
+        if layer is not None:
+            X_batch = adata_subset.layers[layer]
+        else:
+            X_batch = adata_subset.X
+
+        # Prepare clipping
+        if clip is None:
+            n = X_batch.shape[0]
+            clip = np.sqrt(n)
+        if clip < 0:
+            raise ValueError("Pearson residuals require `clip>=0` or `clip=None`.")
+
+        if sp_sparse.issparse(X_batch):
+            sums_genes = np.sum(X_batch, axis=0)
+            sums_cells = np.sum(X_batch, axis=1)
+            sum_total = np.sum(sums_genes).squeeze()
+        else:
+            sums_genes = np.sum(X_batch, axis=0, keepdims=True)
+            sums_cells = np.sum(X_batch, axis=1, keepdims=True)
+            sum_total = np.sum(sums_genes)
+
+        # Compute pearson residuals in chunks
+        residual_gene_var = np.empty((X_batch.shape[1]))
+        for start in np.arange(0, X_batch.shape[1], chunksize):
+            stop = start + chunksize
+            mu = np.array(sums_cells @ sums_genes[:, start:stop] / sum_total)
+            X_dense = X_batch[:, start:stop].toarray()
+            residuals = (X_dense - mu) / np.sqrt(mu + mu ** 2 / theta)
+            residuals = np.clip(residuals, a_min=-clip, a_max=clip)
+            residual_gene_var[start:stop] = np.var(residuals, axis=0)
+
+        # Add 0 values for genes that were filtered out
+        unmasked_residual_gene_var = np.zeros(len(nonzero_genes))
+        unmasked_residual_gene_var[nonzero_genes] = residual_gene_var
+        residual_gene_vars.append(unmasked_residual_gene_var.reshape(1, -1))
+
+    residual_gene_vars = np.concatenate(residual_gene_vars, axis=0)
+
+    # Get rank per gene within each batch
+    # argsort twice gives ranks, small rank means most variable
+    ranks_residual_var = np.argsort(np.argsort(-residual_gene_vars, axis=1), axis=1)
+    ranks_residual_var = ranks_residual_var.astype(np.float32)
+    # count in how many batches a genes was among the n_top_genes
+    highly_variable_nbatches = np.sum(
+        (ranks_residual_var < n_top_genes).astype(int), axis=0
+    )
+    # set non-top genes within each batch to nan
+    ranks_residual_var[ranks_residual_var >= n_top_genes] = np.nan
+    ranks_masked_array = np.ma.masked_invalid(ranks_residual_var)
+    # Median rank across batches, ignoring batches in which gene was not selected
+    medianrank_residual_var = np.ma.median(ranks_masked_array, axis=0).filled(np.nan)
```

Comment:
I agree that exposing multiple options could make sense here. Any opinion @gokceneraslan and  @adamgayoso which ones we should implement / which one should be the default? 

I just copied this part from the seurat_v3 batch integration, so I don't have an expert opinion - but if I remember correctly, the other flavors use `dispersions_norm` to break ties instead of using the median rank. If I read the code correctly, this is the within-batch dispersion averaged over batches.

We could also do something similar here: Compute residual variance within each batch, take the average across batches, and use that to break ties instead of using median rank.

Reply:
I suggest to discuss this in new issue as this is out of scope of this PR